src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -158,6 +222,140 @@ async def generate_lockfile(
         ),
     )
 
+
+@rule
+async def generate_updated_lockfile(
+    update_req: UpdatePythonLockfileRequest,
+    generate_lockfiles_subsystem: GenerateLockfilesSubsystem,
+    pex_lock_subsystem: PexLockSubsystem,
+    python_setup: PythonSetup,
+) -> ProcessResult:
+    req = update_req.gen_req
+    resolve_config = await Get(ResolvePexConfig, ResolvePexConfigRequest(req.resolve_name))
+
+    for project in pex_lock_subsystem.project:
+        if PipRequirement.parse(project).project_name != project:
+            raise ValueError(
+                f"project {project} is not a bare name; specifier, markers must be specified in Pants targets"
```

Comment:
I'm not sure I understand what the means, could you give an example?

Reply:
I know what it means, but Pex itself doesn't have this restriction. Why does Pants need it? Every time Pants gets between a user and a tool and meddles is a new chance to infuriate.

Reply:
So the complication here is where the authoritative list of projects and their constraints lives.  In pants today `libfoo<2` might be in a requirements.txt file, a poetry style file, in BUILD files, or split between any of those places and a constraints.txt file.  If on the cli someone did `--project=libfoo>3`, but had `libfoo<2` in requirements.txt` what should we do?
 * Take the versions in vcs as authoritative? (Current PR)
 * Take the cli as authoritative and leave the vcs file ambiguity?
 * Some permutation of are-you-sure or warnings?
 * Develop concrete syntax trees for editing all of these config files, sort of along the lines of #12880?
 
I think maintaining consistency between the requirements and lockfile is the "least surprising" behavior, but reorganize there are no obviously right expectations.

As I understand them (and maybe I'm confused; I agree having each tool have slightly different semantics makes it hard to follow), the semantics of `pex lock update --project=libfoo` is:
 * If libfoo is not present, use the latest version
 * If libfoo is present in the lock's `requirements`, *replace* the requirements and use the latest libfoo
 * If libfoo is present in the lock's `constraints`, install the latest version of libfoo that abides by the constraints.
 *  If libfoo is present in the lock's `constraints` and cli version is outside them, use the new version, issue a warning, and update the constraints.
 
I think this makes a lot of sense for a tool that is used to do all sorts of spiffy sometimes ad-hoc things and not narrowly focused on doing requirements-->lock a'-la `poetry` or Pants.
 
As a not fully formed feature request, if PEX had a way to say "try to update `libfoo` within the bounds of the `requirements` and `constraints` of the lockfile" and be happy to "get out of the way" and use that.   (FWIW I *think* this is what [poetry does with a similar subcmd](https://python-poetry.org/docs/cli/#update) but that's not a tool I've used before.)

A related PEX feature might be a way of permuting the constraints without changing the requirements.

Reply:
@cburroughs maybe a good question, but the best 1st stop is to ask the tool:
```
$ pex3 lock create "ansicolors<1.1.8" -o lock.json
$ pex3 lock update -p "ansicolors>1.1.8" lock.json
ERROR: The following lock update constraints could not be satisfied:
ansicolors>1.1.8
Encountered 1 error updating lock.json:
1.) cp312-cp312-manylinux_2_35_x86_64: pid 87643 -> /home/jsirois/.pex/venvs/344fa41f0c33753264e54bcca2d4d078936e247e/0de1795ad4486f45ee94cecd983e9905b6b11dc9/bin/python -sE /home/jsirois/.pex/venvs/344fa41f0c33753264e54bcca2d4d078936e247e/0de1795ad4486f45ee94cecd983e9905b6b11dc9/pex --disable-pip-version-check --no-python-version-warning --exists-action a --no-input --isolated -q --cache-dir /home/jsirois/.pex/pip/23.2/pip_cache --log /tmp/pex-pip-log.nod0b_7e/pip.log download --dest /tmp/tmp5vzcj7a3/home.jsirois.bin.pex.venv.bin.python3.12 --constraint /tmp/lock_update.10b2bohk.constraints.txt ansicolors<1.1.8 --retries 5 --timeout 15 exited with 1 and STDERR:
ERROR: Cannot install ansicolors<1.1.8 because these package versions have conflicting dependencies.
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

 The conflict is caused by:
     The user requested ansicolors<1.1.8
     The user requested (constraint) ansicolors>1.1.8

 To fix this you could try to:
 1. loosen the range of package versions you've specified
 2. remove package versions to allow pip attempt to solve the dependency conflict
```

Now, that's an error from Pip behind the scenes and may be slightly confusing, but.

Reply:
So I'd amend my reply above to just say, make sure you totally understand the tool your integrating with Pants before deciding to do extra stuff that may get between the user and the tool. When I worked on Pants, I preferred not using Pants _at all_ to develop an integration, but complete the work on the command line. Then, once sorted, and the tool behavior was totally sussed, get Pants to run those same Processes.

Reply:
FWIW, the place I've found constrained updates useful is to downgrade a security vulnerability when no fix has yet been released, or has, but is otherwise unusable for other reasons. Now, I did this using cargo and not Pex, but the scenario seems both generic and likely.

Reply:
> As I understand them (and maybe I'm confused; I agree having each tool have slightly different semantics makes it hard to follow), the semantics of pex lock update --project=libfoo is:

Better than semantics, here's what Pex actually does:
1. Take the list of requirements as-is from the lock file being updated.
2. Take the list of constraints as-is from the lock file being updated.
3. Modify the constraints list to pin all projects not specified by `-p <req>` updates, replace and constraints that overlap with the `-p <req>` with those reqs, and add any non-overlapping reqs to constraints if they correspond to a project in the lock or else requirements if they are a new project not in the lock.
4. Call `pip download <reqs> --constraint <constraints> ...`.

Reply:
Thanks for the fulsome response, advice and context.  I admit I optimistically though this would be a relatively minor extension to how Pants uses used Pex and didn't spend a lot of time building intuition with `pex3 lock update` in isolation.  However, I think my description of what the tool does still matches what I observe when I ask the tool, at least on this point:

>    If libfoo is present in the lock's requirements, replace the requirements and use the latest libfoo
 
 ```
$ pex3 lock create "ansicolors<1.1.7" -o lock.json
$ pex3 lock update -p "ansicolors" lock.json
Updates for lock generated by cp311-cp311-manylinux_2_37_x86_64
  Updated ansicolors from 1.1.6 to 1.1.8
```

So a bare `--project=name` uses the latest version available, not the latest version given under `requirements`. 

Your example has `ansicolors>1.1.8` but the latest version on PyPi today *is* `1.1.8`:
```
$ pex3 lock create "ansicolors>1.1.8" -o lock.json
pid 32669 -> /home/ecsb/.pex/venvs/5dfab84ad94b45f66204b4c844b864f63a0b1cbc/73c7d7c618fac01449fa6e57a21210c0eb8a8f21/bin/python -sE /home/ecsb/.pex/venvs/5dfab84ad94b45f66204b4c844b864f63a0b1cbc/73c7d7c618fac01449fa6e57a21210c0eb8a8f21/pex --disable-pip-version-check --no-python-version-warning --exists-action a --no-input --use-deprecated legacy-resolver --isolated -q --cache-dir /home/ecsb/.pex/pip/20.3.4-patched/pip_cache --log /tmp/pex-pip-log.i2fhhg78/pip.log download --dest /tmp/tmpkldn460x/usr.bin.python3.11 ansicolors>1.1.8 --retries 5 --timeout 15 exited with 1 and STDERR:
ERROR: Could not find a version that satisfies the requirement ansicolors>1.1.8 (from versions: 1.0, 1.0.1, 1.0.2, 1.1.5, 1.1.6, 1.1.7, 1.1.8)
ERROR: No matching distribution found for ansicolors>1.1.8
```

Reply:
I deny your example:
```
$ pex3 lock create "ansicolors<1.1.7" -o lock.json
$ pex3 lock update -p "ansicolors" lock.json
Updates for lock generated by cp312-cp312-manylinux_2_35_x86_64
  There were no updates for ansicolors
$ pex3 lock export lock.json
ansicolors==1.1.6 \
  --hash=sha256:e471e2364ecf78bd01868f9ad05b036aad3f572cba37278a3126d8b460812756
```

@cburroughs what version of `pex` are you using in your example above?

Reply:
```
$ pex3 --version
2.1.159
```

(I get the same behavior running from source at `58c6e8e1832ad8f871daf0439aed49388f8aa942`.)

Reply:
Aha, you're on python3.11, and with your concise argument set, you get legacy Pip (20.3.4) + legacy resolver. I'm on 3.12 which has a Pip 23.2 floor and Pip 23.2 no longer even allows the legacy resolver. You'll find you get the behavior I showed if you add the `--resolver-version pip-2020-resolver` option, which is what Pants does unconditionally. In short, you were observing the old broken Pip resolver which was well known not to respect requirements / constraints!

Reply:
So, in your CLI testing, always include `--resolver-version pip-2020-resolver` at least to get apples-apples with what Pants will see. I think you can mostly get away with leaving the other Pants options (`--interpreter-constraint` and `--target-system *`) off in your testing.

Reply:
Matching behavior confirmed with `--resolver-version pip-2020-resolver`.  Thank you.

Then if I've un-counfounded my understanding, the ability to change the requirements (or constraints) of an existing project would be a new feature request, correct?  I see that #15704 as originally requested was narrowly about adding *new* projects, while #12880 and #15275 had somewhat broader in discussion.

(The motivation is that we have > 120 project requirements.  For many of them like (ex: `types-foo`) there is little value in attaching a version specifier, while some important or large dependencies are  `==` pinned to a specific version.  I would like to update `important_package==1.2.3` to `important_package==1.2.4` without pulling in unrelated changes in the transitive dependency graph.)

Reply:
I don't care about feature requests so much - I don't use Pants. I care about not getting bugged because Pants made the choice to be more restrictive than it need be.

If the written down requirement is, say, `requests>=2,<3` and requests 2.3.4 is locked and a newly reported sec vuln in 2.3.4 can be solved by `-p requests<2.3.4` using Pex straight up (i.e.: upgrading does _not_ solve it) but Pants says no, that would be infuriating, because that works today with Pex locks. You can move an existing locked requirement anywhere in its written down requirement range as long as the transitive resolve leads to no conflict.

FWICT the experiments above show that Pex will fail to lock anything not in the written down requirement range (using the pip-2020-resolver); so its 0 skin off your back to just allow -p like Pex does. If the user enters some conflicting `-p requests>3`, they get yelled at immediately and are forced to go edit the orignal written down `requests>=2,<3` controlled by Pants and all is good.

Reply:
FWIW, and this PR is not settled yet, but if it were settled on nannying Pex's `-p` un-necessarily, that would be exactly in one of the class of reasons I don't use Pants. It babysits too much for my taste by half.

Reply:
Thank you again for the response and your work on Pex. I find it useful both through Pants as a standalone tool.  I don't have a strongly held opinion on when Pants is net helpful by imposing additional checks on what is passed to underlying tools.  I understand you lean strongly against doing so!  And those reasons make sense to me.

> the experiments above show that Pex will fail to lock anything not in the written down requirement range (using the pip-2020-resolver); so its 0 skin off your back to just allow -p like Pex does.

Yes, I'm inclined towards choices that decease the number of `if` statements in this PR.  The check was added when my understanding of the behavior was muddled by the old resolver as you pointed out.

> If the user enters some conflicting -p requests>3, they get yelled at immediately and are forced to go edit the original written down requests>=2,<3 controlled by Pants and all is good.

But after being so yelled at and editing the original to new bounds, `lock update` can't then do that update right? There is no `--project-with-new-version-specifier` and as you have pointed out `--project` will maintain the original specifier.  I don't think this is a Pants specific thing.  A project using the  common Python pattern of having a `requirements.txt` file alongside a lockfile would be in the same boat when making changes to said `requirements.txt`.

Just like "add a new project without changing other entries" or "remove a project without changing other entries" is reasonable, I think  "change the version specifier of one project without changing other entries" is for the same reasons.

I'm happy to take that to another issue if you would prefer.  I understand you have no obligation to support that use case.

Reply:
> But after being so yelled at and editing the original to new bounds, lock update can't then do that update right? 

Correct - you need to re-gen the lock. The current semantics of `pex3 lock update` are to update the original lock given all its settings. You can only add new requirements or push original requirements around in their version range as originally written down.

> Just like "add a new project without changing other entries" or "remove a project without changing other entries" is reasonable, I think "change the version specifier of one project without changing other entries" is for the same reasons.

I agree. If you file an issue for these two - remove and replace (instead of the current adjust) - I can get those knocked out tomorrow.

Reply:
This is now available as `-R <project name>` / `--replace-project <project name>` (#2335) released in [Pex 2.1.160](https://github.com/pantsbuild/pex/releases/tag/v2.1.160).